vendor/k8s.io/kubernetes/pkg/kubectl/resource_filter.go

```diff
@@ -0,0 +1,100 @@
+package kubectl
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type filterEntry struct {
+	filterFunc reflect.Value
+}
+
+// FilterOptions implements PrintOptions
+type FilterOptions struct {
+	PrintOptions
+}
+
+type ResourceFilter struct {
+	filterMap map[reflect.Type]*filterEntry
+	options   FilterOptions
+}
+
+func NewResourceFilter(opts *FilterOptions) *ResourceFilter {
+	filter := &ResourceFilter{
+		filterMap: make(map[reflect.Type]*filterEntry),
+		options:   *opts,
+	}
+
+	filter.addDefaultHandlers()
+	return filter
+}
+
+func (f *ResourceFilter) addDefaultHandlers() {
+	f.Handler(filterPod)
+}
+
+// Handler adds a filter handler to a ResourceFilter instance.
+// See validateFilterFunc for required method signature.
+func (f *ResourceFilter) Handler(filterFunc interface{}) error {
+	filterFuncValue := reflect.ValueOf(filterFunc)
+	if err := f.validateFilterFunc(filterFuncValue); err != nil {
+		glog.Errorf("Unable to add print handler: %v", err)
+		return err
+	}
+	objType := filterFuncValue.Type().In(0)
+	f.filterMap[objType] = &filterEntry{
+		filterFunc: filterFuncValue,
+	}
+	return nil
+}
+
+// validateFilterFunc validates the filter handler signature.
+// filterFunc is the function that will be called to filter an object.
+// It must be of the following type:
+//  func filterFunc(object ObjectType, options FilterOptions) error
+// where ObjectType is the type of the object that will be printed.
+func (f *ResourceFilter) validateFilterFunc(filterFunc reflect.Value) error {
```

Comment:
> Why is this reflection? You can filter on runtime.Object

It is to validate a "handler" function. I copied this from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource_printer.go#L415. My thought process was that if this were to be used on resources other than pods in the future, each would require a different function that would know how to filter than specific resource type (kind of like `resource_printer.go`).